File: Examenes-EDD-6/2016-Febrero/Enunciado/Haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name: Paco Molina Cuenca
-- Student's group: Doble Grado Matemáticas e Ingeniería Informática
--
-- Data Structures. Escuela Térnica Superior de Ingenieria Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, remove, extractCycle, connectCycles, vertexInCommon, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian = undefined

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove = undefined

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle = undefined
   
-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles = undefined

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon = undefined

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle = undefined

File: Examenes-EDD-7/2016-Febrero/Enunciado/Java/dataStructures/graph/EulerianCycle.java
/**
 * Student's name: Ángel Manuel Soria Gil
 * Student's group: Doble Grado Matemáticas e Ingeniería Informática
 * Data Structures. E.T.S.I. Informática. UMA.
 */

package dataStructures.graph;
import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.set.*;
import dataStructures.list.*;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {

    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {

    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {

    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {

    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {

    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {

    }
}
